### ISSUE TITLE
Delta programming (restrict uploads to changes only)

### ISSUE THREAD

**Issue by navnavnav** [bloombloombloom/Bloom#95]
Currently, when programming the target, we upload the entire program to the target. 

But how often does the entire program change between uploads? I think it's safe to say that in most cases, only relatively small changes are made between uploads. This means that we're wasting time updating parts of program memory that hasn't changed.

I believe we could drastically improve upload performance by generating a diff of the current binary and the new binary provided by GDB, and then only applying the changes to program memory whilst ignoring the rest.

First few points that come to mind:

- We could make use of the program memory cache as a means to keep track of changes
- The first upload for each debug session will still be slow, as the program memory cache will not be populated at that point
- We'll need a way to disable/ignore GDB's erase packet. GDB sends an erase packet to Bloom at the beginning of each programming session. Servicing this would mean the entire program memory would need to be updated, rendering this feature useless.
- This feature will make little difference for IDE users. At least for the IDEs that restart Bloom whenever the user wishes to rebuild the program (as CLion currently does). I've already [brought this up with JetBrains](https://youtrack.jetbrains.com/issue/CPP-35677/Add-option-to-maintain-active-GDB-connection-when-rebuilding-and-rerunning-embedded-program-for-better-faster-debugging), so it might be worth putting this on hold until they address that issue.
- This feature will not be suitable for programs that can alter the target's program memory (such as bootloaders). It should be optional

More to follow

**Comment by navnavnav** [comment]
Thinking about this some more: This will only work for debugWire targets, because there's no requirement for a chip/section erase before programming. For other targets (PDI, UPDI and JTAG), we must perform a chip/section erase before we can program the target. So we'll have to write the whole program binary for every programming session.

This is not worth the effort if it only works on debugWire targets.

I'm going to leave this open for the time-being, as it may be appropriate for a new target family I'm looking to support at some point.

**Comment by navnavnav** [comment]
This is appropriate for the RISC-V targets that I plan to support in the next feature release (see https://github.com/bloombloombloom/Bloom/issues/96).

With those targets, we're able to perform partial page writes to program memory without having to erase the entire chip. This means we can perform 'delta programming' on these targets.

Not sure when I'll get to this, or if it will be included in the next release, with the initial RISC-V support. Will keep the ticket updated,

**Comment by navnavnav** [comment]
Quick update:

Have made good progress with this. Have got delta programming working with WCH RISC-V targets, as well as UPDI and debugWire AVR8 targets.

It's fast! 

On the CH32V003, it takes 1.2 seconds to upload a 12KB program (and reset the target). With delta programming enabled, it takes 110 milliseconds for small changes to the same program. For the smallest change (e.g. changing the value of some argument in a function call, resulting in a single byte difference), it takes just 68 milliseconds.

So that's a 10X+ improvement to programming speeds on WCH RISC-V targets, assuming only small changes to the program have been made.

On the AVR64DU32 Curiosity Nano, it takes 4.2 seconds to upload a 30KB program (including ~300 milliseconds for the resetting of the target). With delta programming enabled, it takes 400 milliseconds for small changes to the same program. 

UPDI targets require page alignment when programming flash. Bloom uses the program memory cache to align the delta segments. And if the cache doesn't contain the data required for alignment, we'll just fill the page with `0xFF` bytes. This means we do not have to perform additional read operations for the purposes of alignment. This saves a good 200-300 milliseconds, in some cases.

So again, that's a 10X improvement to programming speeds on UPDI targets.

On the ATmega328PB (Xplained Mini), it takes 54 seconds to upload a 20KB program. Yikes! That includes 300-400 milliseconds for the reset. With delta programming enabled, it takes 820 milliseconds for small changes to the same program. Keep in mind that about half of that is due to resetting the target...

So that's a 60X+ improvement. YIKES

Looking forward to shipping this with v2.0.0, which will likely be around beginning-to-mid February.

**Comment by navnavnav** [comment]
Development work for this is complete. Took longer than expected because testing delta programming had revealed quite a few bugs in the WCH-Link driver and firmware.

These changes will be shipped with v2.0.0. Will close this ticket then.

**Comment by navnavnav** [comment]
These changes have been included in v2.0.0.
